massa-execution/src/worker.rs

```diff
@@ -255,106 +253,137 @@ impl ExecutionWorker {
-                    self.last_final_slot = next_final_slot;
-                } else {
-                    self.ordered_pending_css_final_blocks.push((b_id, block));
-                    break;
+                    // there are no more CSS-final blocks
+                    None => break,
                 }
+
+                cur_slot = cur_slot.get_next_slot(self.cfg.thread_count)?;
             }
         }
 
-        // list remaining CSS finals + new blockclique
-        self.ordered_active_blocks = self
-            .ordered_pending_css_final_blocks
+        // process SCE-active blocks
+        let mut sce_active_blocks: BTreeMap<Slot, (&BlockId, &Block)> = blockclique
             .iter()
-            .cloned()
+            .filter_map(|(b_id, b)| {
+                if b.header.content.slot <= self.last_final_slot {
+                    // eliminate blocks that are not from a stricly later slot than the current latest SCE-final one
+                    return None;
+                }
+                Some((b.header.content.slot, (b_id, b)))
+            })
             .chain(
-                blockclique
-                    .into_iter()
-                    .filter(|(_b_id, b)| b.header.content.slot > self.last_final_slot),
+                self.pending_css_final_blocks
+                    .iter()
+                    .map(|(k, (b_id, b))| (*k, (b_id, b))),
             )
             .collect();
+        if let Some(max_css_active_slot) = sce_active_blocks.last_key_value().map(|(s, _v)| *s) {
+            let mut cur_slot = self.last_final_slot.get_next_slot(self.cfg.thread_count)?;
+            while cur_slot <= max_css_active_slot {
+                match sce_active_blocks.first_key_value().map(|(s, _v)| *s) {
+                    // there is a CSS-active block at cur_slot
+                    Some(b_slot) if b_slot == cur_slot => {
+                        // remove the entry from sce_active_blocks (cannot panic, checked above)
+                        let (_b_slot, (b_id, block)) = sce_active_blocks.pop_first().unwrap();
+                        // execute the block as a SCE-active VM step
+                        self.push_request(ExecutionRequest::RunActiveStep(ExecutionStep {
+                            slot: cur_slot,
+                            block: Some((*b_id, block.clone())),
+                        }));
+                        // update cursor
+                        self.last_active_slot = cur_slot;
+                    }
 
-        // sort active blocks
-        self.ordered_active_blocks
-            .sort_unstable_by_key(|(_b_id, b)| b.header.content.slot);
+                    // there is no CSS-active block at cur_slot
+                    Some(_b_slot) => {
+                        // execute a miss as an SCE-active VM step
+                        self.push_request(ExecutionRequest::RunActiveStep(ExecutionStep {
+                            slot: cur_slot,
+                            block: None,
```

Comment:
could you not simply do this with one call to `push_request`, where `block` would be an option acquired earlier?

Reply:
I didn't do it for 3 reasons:
* it makes the spec clearer
* it better matches the spec
* there is a third "None => break," case that would make the split heavier

Basically it has no impact on performance and makes the code more explicit. Duplication is very minimal.

Reply:
Ok